pyirf/io/gadf.py

```diff
@@ -303,3 +309,201 @@ def create_rad_max_hdu(
+
+    cuts = list()
+    for file_name in files:
+        table = QTable.read(file_name, hdu=extname)
+        cuts.append(table)
+    return cuts
+
+
+def read_fits_bins_lo_hi(file_name, extname, tags):
+    """
+    Reads from a fits file two arrays of tag_LO and tag_HI.
+    If more then one file is given it checks that all of them are consistent before returning the value
+
+    Parameters
+    ----------
+    file_name: string or list of string
+        file to be read, if a list of
+    extname: string
+        name of the extension to read the data from in fits file
+    tags: list of string
+        names of the field in the extension to extract, _LO and _HI will be added
+
+    Returns
+    -------
+    bins: list of astropy.units.Quantity
+        list of ntuples (LO, HI) of bins (with size of extnames)
+    """
+
+    # to allow the function work on either single file or a list of files convert a single string into a list
+    if isinstance(file_name, str):
+        file_name = [file_name]
+
+    # same to allow to run over single tag
+    if isinstance(tags, str):
+        tags = [tags]
+
+    tags_lo_hi = [(tag + '_LO', tag + '_HI') for tag in tags]
+
+    old_table = None
+    bins = list()
+    for this_file in file_name:
+        table = QTable.read(this_file, hdu=extname)
+        for tag_lo, tag_hi in tags_lo_hi:
+            if old_table is not None:
+                if not old_table[tag_lo][0].shape == table[tag_lo][0].shape:
+                    raise ValueError('Non matching bins in ' + extname)
+                if not ((old_table[tag_lo][0] == table[tag_lo][0]).all() and (old_table[tag_hi][0] == table[tag_hi][0]).all()):
+                    raise ValueError('Non matching bins in ' + extname)
+            else:
+                bins.append([table[tag_lo][0], table[tag_hi][0]])
+        old_table = table
+
+    return bins
+
+
+def read_irf_grid(files, extname, field_name):
+    """
+    Reads in a grid of IRFs for a bunch of different parameters and stores them in lists
+
+    Parameters
+    ----------
+    files: string or list of strings
+        files to be read
+    extname: string
+        name of the extension to read the data from in fits file
+    field_name: string
+        name of the field in the extension to extract
+
+    Returns
+    -------
+    irfs_all: np.array
+        array of IRFs, first axis specify the file number(if more then one is given),
+        second axis is the offset angle, rest of the axes are IRF-specific
+    theta_bins: astropy.units.Quantity[angle]
+        theta bins
+    """
+
+    # to allow the function work on either single file or a list of files convert a single string into a list
+    if isinstance(files, str):
+        files = [files]
+
+    n_files = len(files)
+
+    irfs_all = list()
+
+    for this_file in files:
+        # [0] because there the IRFs are written as a single row of the table
+        irfs_all.append(QTable.read(this_file, hdu=extname)[field_name][0])
```

Comment:
The transpose is missing here, this is why you have to do non-trivial transposes in edisp.

This goes back to the point that I really don't understand why we need this `read_irf_grid` function. Simple functions that read a table once and return the arrays. And stacking can then just be `edisp = np.stack(edisps)`

Reply:
Also, keeping the table around is good, since it contains the metadata. So this function is just not as general as you might think it is.

Reply:
I switched to np.stack. 
Please note @maxnoe that the function is also making the required transposition that is there for all the IRFs and extracting the 0-th row - all the operations to convert between the FITS format and the one used inside pyirf. And it is working fine on either individual files or lists of files so I think it is very useful and it makes the custom functions that read individual IRFs simpler. 

I agree that the table includes additional important information via metadata, but I think it would be more useful to extract this metadata and return such concrete information. Nevertheless, first we would need to agree (in #126 ) what information should be put there, with what names, etc.

Reply:
@jsitarek 

> Nevertheless, first we would need to agree (in #126 ) what information should be put there, with what names, etc.

Actually no. Since at no point pyirf will use this metadata, users of pyirf can fill and read any metadata they like and use it for whatever purpose. 

That's also the problem with this `read_irf_grid` function. You would have to open all the files again to also read the metadata.

So, please, write functions that read each IRF type, validating it's the correct type by looking at HDUCLASX header keywords and returning also the metadata.

This functions can than easily be called for multiple files and the results stacked / compared.

I think this has much better advantages than the little code repetition you avoid by  using this read_irf_grid function.

Reply:
> @jsitarek
> 
> > Nevertheless, first we would need to agree (in #126 ) what information should be put there, with what names, etc.
> 
> Actually no. Since at no point pyirf will use this metadata, users of pyirf can fill and read any metadata they like and use it for whatever purpose.
> 
well, I would not be so sure about this. If we really have standardized metadata we can use them inside pyirf to do e.g. standarized interpolation over those values. 

> That's also the problem with this `read_irf_grid` function. You would have to open all the files again to also read the metadata.
> 
as I explained above the metadata processing can be also added to this function. Please note that they should be likely the same for all the types of IRFs (no matter if you want to read Aeff or energy migration you still want to know at which zenith it was produced) so it takes perfect sense to read them in standarized way

> So, please, write functions that read each IRF type, validating it's the correct type by looking at HDUCLASX header keywords and returning also the metadata.
> 
I just implemented checking of all the HDUCLASX headers. 
Again because of using one universal function it was pretty easy, especially that many of those fields are the same for different IRFs.


> This functions can then easily be called for multiple files and the results stacked / compared.
> 
> I think this has much better advantages than the little code repetition you avoid by using this read_irf_grid function.

With every thing that we discuss in this thread the amount of code repetition in the approach that you suggested would grow larger and larger. Already now we have 3 issues that are repeated for all IRFs: 
- swapping of axis sequence 
- checks of  HDUCLASX headers
- combining IRFs produced for different parameters

and we have two IRFs already implemented, and a few more that can be added easily. So in the end I think we are avoiding a lot of code repetition